### ISSUE TITLE
Out of memory kill event after meryl

### ISSUE THREAD

**Issue by AdamStuckert** [marbl/canu#1757]
Hi, 

I'm having a memory issue that occurs immediately after meryl finishes. This is similar to a couple other issues (e.g. https://github.com/marbl/canu/issues/1561), but upping executiveMemory has not solved it for me. It does submit a job (called `canu_parvus` for me), but this fails with the memory issue after a few seconds. I'm running this from a Linux cluster which uses SLURM.

My command:

```
canu \
-p parvus -d canu \
genomeSize=6g \
correctedErrorRate=0.055 \
gnuplot=undef \
gridOptions="--mem-per-cpu=10g" \
executiveMemory=64 \
-pacbio raw_data/S_parvus_smrtcell_1.fasta \
-pacbio raw_data/S_parvus_smrtcell_2.fasta \
-pacbio raw_data/S_parvus_smrtcell_3.fasta
```

Log output:
```
Found perl:
   /mnt/lustre/macmaneslab/ams1236/Oyster_River_Protocol/software/anaconda/install/bin/perl
   This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-linux-thread-multi

Found java:
   /usr/bin/java
   openjdk version "1.8.0_232"

Found canu:
   /mnt/oldhome/macmaneslab/ams1236/software/canu-2.0/Linux-amd64/bin/canu
   Canu 2.0

-- Canu 2.0
--
-- Detected Java(TM) Runtime Environment '1.8.0_232' (from 'java') with -d64 support.
-- No path to gnuplot executable.  Plots disabled.
-- Detected 64 CPUs and 995 gigabytes of memory.
-- Detected Slurm with 'sinfo' binary in /usr/bin/sinfo.
-- Detected Slurm with task IDs up to 19999 allowed.
--
-- Found   4 hosts with  40 cores and  375 GB memory under Slurm control.
-- Found   4 hosts with  24 cores and  502 GB memory under Slurm control.
-- Found   6 hosts with  40 cores and  753 GB memory under Slurm control.
-- Found   2 hosts with  64 cores and  994 GB memory under Slurm control.
-- Found   1 host  with  40 cores and  186 GB memory under Slurm control.
-- Found  27 hosts with  24 cores and  124 GB memory under Slurm control.
-- Found   4 hosts with  24 cores and  503 GB memory under Slurm control.
--
--                         (tag)Threads
--                (tag)Memory         |
--        (tag)             |         |  algorithm
--        -------  ----------  --------  -----------------------------
-- Grid:  meryl     37.000 GB    8 CPUs  (k-mer counting)
-- Grid:  hap       16.000 GB   20 CPUs  (read-to-haplotype assignment)
-- Grid:  cormhap   37.000 GB    8 CPUs  (overlap detection with mhap)
-- Grid:  obtovl    24.000 GB    8 CPUs  (overlap detection)
-- Grid:  utgovl    24.000 GB    8 CPUs  (overlap detection)
-- Grid:  cor       24.000 GB    4 CPUs  (read correction)
-- Grid:  ovb        4.000 GB    1 CPU   (overlap store bucketizer)
-- Grid:  ovs       32.000 GB    1 CPU   (overlap store sorting)
-- Grid:  red       37.000 GB    8 CPUs  (read error detection)
-- Grid:  oea        8.000 GB    1 CPU   (overlap error adjustment)
-- Grid:  bat      994.000 GB   64 CPUs  (contig construction with bogart)
-- Grid:  cns        -.--- GB    8 CPUs  (consensus)
-- Grid:  gfa      128.000 GB   64 CPUs  (GFA alignment and processing)
--
-- In 'parvus.seqStore', found PacBio CLR reads:
--   PacBio CLR:               3
--
--   Raw:                      1
--
-- Generating assembly 'parvus' in '/mnt/oldhome/macmaneslab/ams1236/footloose_genome/canu':
--    - correct raw reads.
--    - trim corrected reads.
--    - assemble corrected and trimmed reads.
--
-- Parameters:
--
--  genomeSize        6000000000
--
--  Overlap Generation Limits:
--    corOvlErrorRate 0.2400 ( 24.00%)
--    obtOvlErrorRate 0.0550 (  5.50%)
--    utgOvlErrorRate 0.0550 (  5.50%)
--
--  Overlap Processing Limits:
--    corErrorRate    0.3000 ( 30.00%)
--    obtErrorRate    0.0550 (  5.50%)
--    utgErrorRate    0.0550 (  5.50%)
--    cnsErrorRate    0.0550 (  5.50%)
--
--
-- BEGIN CORRECTION
--
-- No change in report.
-- Meryl finished successfully.  Kmer frequency histogram:
/var/spool/slurmd/job101166/slurm_script: line 48: 57919 Killed                  /usr/bin/env perl $bin/canu -p 'parvus' 'genomeSize=6g' 'correctedErrorRate=0.055' 'gnuplot=undef' 'gridOptions=--mem-per-cpu=10g' 'executiveMemory=64' -pacbio '/mnt/oldhome/macmaneslab/ams1236/footloose_genome/raw_data/S_parvus_smrtcell_1.fasta' -pacbio '/mnt/oldhome/macmaneslab/ams1236/footloose_genome/raw_data/S_parvus_smrtcell_2.fasta' -pacbio '/mnt/oldhome/macmaneslab/ams1236/footloose_genome/raw_data/S_parvus_smrtcell_3.fasta' canuIteration=0
slurmstepd: error: Detected 1 oom-kill event(s) in step 101166.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.
```

Thanks in advance!

**Comment by skoren** [comment]
I think this may be a conflict between your parameters. You specify both `gridOptions="--mem-per-cpu=10g"` and `executiveMemory=64`. The first one will overwrite the second one and since the executive uses 1 core, it will be limited to 10gb. You can confirm this by checking the submit command in the last `canu-scripts/canu.*.out` file before the one with the error. 

Looking at the configuration, 10gb should be safe for most of your steps (except bogart) but Canu will request the appropriate memory anyway so I'd see what happens if you removed `gridOptions="--mem-per-cpu=10g"`. You could also try increasing it to 15 or 20.

**Comment by AdamStuckert** [comment]
Thanks for responding. Neither `executiveMemory=64` nor `--mem-per-cpu=20` flags alone yield any change. It submits and runs out of memory in < 10 seconds of runtime with the same error as above in both cases. 

Submission 1:

```
canu -p parvus -d canu genomeSize=6g correctedErrorRate=0.055 gnuplot=undef executiveMemory=64 -pacbio raw_data/S_parvus_smrtcell_1.fasta -pacbio raw_data/S_parvus_smrtcell_2.fasta -pacbio raw_data/S_parvus_smrtcell_3.fasta
```

Submission 2:
```
canu -p parvus -d canu genomeSize=6g correctedErrorRate=0.055 gnuplot=undef gridOptions="--mem-per-cpu=20g" -pacbio raw_data/S_parvus_smrtcell_1.fasta -pacbio raw_data/S_parvus_smrtcell_2.fasta -pacbio raw_data/S_parvus_smrtcell_3.fasta
```

**Comment by skoren** [comment]
What is the submit command in both those cases? What are the files in ```*/0-*``` in your assembly folder?

**Comment by AdamStuckert** [comment]
Submit commands, in order from above:

```
    cd /mnt/oldhome/macmaneslab/ams1236/footloose_genome/canu
    sbatch \
      --cpus-per-task=1 \
      --mem-per-cpu=64g   \
      -D `pwd` \
      -J 'canu_parvus' \
      -o canu-scripts/canu.09.out  canu-scripts/canu.09.sh
Submitted batch job 101175
```
and
```
    cd /mnt/oldhome/macmaneslab/ams1236/footloose_genome/canu
    sbatch \
      --cpus-per-task=1 \
      --mem-per-cpu=4g \
      --mem-per-cpu=20g  \
      -D `pwd` \
      -J 'canu_parvus' \
      -o canu-scripts/canu.10.out  canu-scripts/canu.10.sh
```

Files in the `*/0-*` folder:

```
-rwxr-xr-x 1 ams1236 macmaneslab 3.5K Jul  8 22:27 meryl-configure.sh
-rw-r--r-- 1 ams1236 macmaneslab 2.7K Jul  8 22:27 parvus.ms16.config.01.out
-rw-r--r-- 1 ams1236 macmaneslab 2.7K Jul  8 22:27 parvus.ms16.config.02.out
-rw-r--r-- 1 ams1236 macmaneslab 2.6K Jul  8 22:27 parvus.ms16.config.04.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.06.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.08.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.12.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.16.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.20.out
-rw-r--r-- 1 ams1236 macmaneslab 2.5K Jul  8 22:27 parvus.ms16.config.24.out
-rw-r--r-- 1 ams1236 macmaneslab 2.4K Jul  8 22:27 parvus.ms16.config.32.out
-rw-r--r-- 1 ams1236 macmaneslab 2.4K Jul  8 22:27 parvus.ms16.config.40.out
-rw-r--r-- 1 ams1236 macmaneslab 2.4K Jul  8 22:27 parvus.ms16.config.48.out
-rw-r--r-- 1 ams1236 macmaneslab 2.3K Jul  8 22:27 parvus.ms16.config.56.out
-rw-r--r-- 1 ams1236 macmaneslab 2.3K Jul  8 22:27 parvus.ms16.config.64.out
-rw-r--r-- 1 ams1236 macmaneslab 2.3K Jul  8 22:27 parvus.ms16.config.96.out
-rwxr-xr-x 1 ams1236 macmaneslab 4.3K Jul  8 22:27 meryl-process.sh
-rwxr-xr-x 1 ams1236 macmaneslab  472 Jul  8 22:27 meryl-make-ignore.pl
-rwxr-xr-x 1 ams1236 macmaneslab 2.3K Jul  8 22:27 meryl-count.sh
-rwxr-xr-x 1 ams1236 macmaneslab  212 Jul  8 22:27 meryl-count.jobSubmit-01.sh
-rw-r--r-- 1 ams1236 macmaneslab   27 Jul  8 22:27 meryl-count.jobSubmit-01.out
-rw-r--r-- 1 ams1236 macmaneslab 4.4K Jul  9 01:08 meryl-count.101161_1.out
-rw-r--r-- 1 ams1236 macmaneslab    0 Jul  9 01:08 meryl-count.success
-rwxr-xr-x 1 ams1236 macmaneslab  218 Jul  9 01:08 meryl-process.jobSubmit-01.sh
-rw-r--r-- 1 ams1236 macmaneslab   27 Jul  9 01:08 meryl-process.jobSubmit-01.out
drwxr-xr-x 2 ams1236 macmaneslab  12K Jul  9 01:09 parvus.ms16
-rw-r--r-- 1 ams1236 macmaneslab 8.5M Jul  9 01:10 parvus.ms16.histogram
-rw-r--r-- 1 ams1236 macmaneslab 8.7M Jul  9 01:10 parvus.ms16.dump
-rw-r--r-- 1 ams1236 macmaneslab 3.8K Jul  9 01:10 meryl-process.101163_1.out
-rw-r--r-- 1 ams1236 macmaneslab 7.3M Jul  9 01:10 parvus.ms16.ignore.gz
```

**Comment by skoren** [comment]
The only thing this step is doing is outputting the histogram, it's possible you have very high copy k-mers which is causing issues (and has been fixed since 2.0). You can just skip this since this is just for display if you touch the file `*/0-*/meryl-process.success` file and then re-start.

**Comment by AdamStuckert** [comment]
Excellent, this solved the issue for me. Thanks very much!
